cdap-common/src/main/java/io/cdap/cdap/common/internal/remote/TaskWorkerHttpHandlerInternal.java

```diff
@@ -120,39 +120,36 @@ public TaskWorkerHttpHandlerInternal(CConfiguration cConf,
     if (enableUserCodeIsolationEnabled) {
       // Run only one request at a time in user code isolation mode.
       this.requestLimit = 1;
-      // Restart the service to clean up and re-claim resources after user code
-      // execution.
-      this.taskCompletionConsumer = (succeeded, taskDetails) -> {
-        taskDetails.emitMetrics(succeeded);
-        runningRequestCount.decrementAndGet();
-        requestProcessedCount.incrementAndGet();
-
-        String className = taskDetails.getClassName();
-
-        if (mustRestart.get()) {
-          stopper.accept(className);
-          return;
-        }
-
-        if (!taskDetails.isTerminateOnComplete() || className == null
-            || killAfterRequestCount <= 0) {
-          // No need to restart.
-          return;
-        }
-
-        if (requestProcessedCount.get() >= killAfterRequestCount) {
-          stopper.accept(className);
-        }
-      };
-
-      enablePeriodicRestart(cConf, stopper);
     } else {
       this.requestLimit = cConf.getInt(TaskWorker.REQUEST_LIMIT);
-      this.taskCompletionConsumer = (succeeded, taskDetails) -> {
-        taskDetails.emitMetrics(succeeded);
-        runningRequestCount.decrementAndGet();
-      };
     }
+
+    // Restart the service to clean up and re-claim resources after user code
+    // execution.
+    this.taskCompletionConsumer = (succeeded, taskDetails) -> {
+      taskDetails.emitMetrics(succeeded);
+      final int pendingRequests = runningRequestCount.decrementAndGet();
+      requestProcessedCount.incrementAndGet();
+
+      String className = taskDetails.getClassName();
+      if (mustRestart.get() && pendingRequests == 0) {
+        stopper.accept(className);
+        return;
+      }
+
+      if (!enableUserCodeIsolationEnabled
```

Comment:
Why do we disable the killAfterRequestCount feature when code isolation is disabled? This is a useful feature

Reply:
I think this is the existing behavior, i.e. not changed in this PR. 

Are you asking in reference to this PR? or a design/implementation level question in general ?

Reply:
Well, we should either change it or change the comments. I believe cdap-defaults does not talk about any dependency.

Reply:
changed the comment in cdap-default.xml

PTAL